### PR TITLE
Swagger api documentation library

### PR DESCRIPTION
**Pull request by freeranger**
PT chore: https://www.pivotaltracker.com/story/show/114070647

Adds Swagger as the documentation library, which is more cross platform and has been donated to the Open API initiative: https://openapis.org

http://swagger.io
https://github.com/richhollis/swagger-docs

to test.

```
$ bundle exec rake swagger:docs
$ rails s
```

then navigate to `http://localhost:3000/api`

You can test out the api calls directly in the documentation but only Ping will return anything "meaningful" until there are users in the database....

screen shot 2016-02-21 at 23 34 35

### CODE REVIEW

File: public/swagger [hunk 1]
```diff
@@ -0,0 +1 @@
+Subproject commit 86be831a8b4ef73ce0a83c94fec93a526994ef95
```
Comment (tochman) [inline public/swagger:1]: What does this setting do?

File: config/initializers/swagger.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  def self.base_api_controller; ApiController end
+
+  def self.transform_path(path, api_version)
+    "apidocs/#{path}"
+  end
+end
+
+host = ENV['IP'] || 'localhost'
+port = ENV['PORT'] || '3000'
+apiUrl = "http://#{host}:#{port}"
```
Comment (tochman) [inline config/initializers/swagger.rb]: try sticking to using `snake_case` for variable names

File: app/controllers/api/v1/users_controller.rb [hunk 3]
```diff
@@ -1,9 +1,26 @@
...
+    response :unauthorized
+  end
+
   def index
     @users = User.all
   end
 
+  swagger_api :show do
+    summary 'Responds with a user resource'
+    param :path, :user, :string, :required, "User Id"
```
Comment (tochman) [inline app/controllers/api/v1/users_controller.rb:17]: Are we requiring the resource `id` as `"1"` or as `1` ? If the latter them we should probably change `:string` to `:integer` or `: fixnum`

File: app/controllers/api/v0/ping_controller.rb [hunk 4]
```diff
@@ -1,5 +1,11 @@
 class Api::V0::PingController < ApiController
 
+  swagger_controller :ping, 'Ping'
+
+  swagger_api :index do
+    summary 'Responds with "Pong"Returns "Pong" to check the api is alive"'
```
Comment (tochman) [inline app/controllers/api/v0/ping_controller.rb:6]: Is there a way to display an example json output?

**Comment by tochman** [comment]
Can you add some screenshots to this PR and some tests?

File: app/controllers/api/v0/ping_controller.rb (hunk 4 above)
Comment (freeranger) [inline app/controllers/api/v0/ping_controller.rb:6]: Run the documentation endpoint and you can test the API directly and see the output there

File: app/controllers/api/v1/users_controller.rb (hunk 3 above)
Comment (freeranger) [inline app/controllers/api/v1/users_controller.rb:17]: Sorry, no idea - I can't tell from the endpoint and there are no cucumber tests for it.
I just added this documentation so you can see what it is like - I thought the idea was to use "friendly" usernames so I assume it should be /api/users/(username)

File: app/controllers/api/v0/ping_controller.rb (hunk 4 above)
Comment (tochman) [inline app/controllers/api/v0/ping_controller.rb:6]: In the documentation I mean

**Comment by tochman** [comment]
As acceptance tests go, I added something like this:

``` gherkin
Feature: As developer of Project Unify API
  In order to make usage of my API easier
  I would like to make my api documentation accessible over internet

  Scenario: Access API documentation
    Given I visit the "api-doc" page
    Then I should see "Project Unify API...."
    And I should see "Some description....."
```

File: public/swagger (hunk 1 above)
Comment (freeranger) [inline public/swagger:1]: The swagger ui is a separate project which we are referencing, not including the source code of - I didn't add that line explicitly (though I did add a sub project reference) but I'm guessing this means we reference a specific commit of the swagger ui project

**Comment by freeranger** [comment]
Can't seem to get a test working for this - maybe because it's all in process now or maybe swagger expects the api to be running or maybe it's just me....

File: .gitignore [hunk 5]
```diff
@@ -14,8 +14,9 @@
 # Ignore all logfiles and tempfiles.
 /log/*
 !/log/.keep
-/tmp
+/tmp                                                                                             rake swagger:docs
```
Comment (tochman) [inline .gitignore:18]: ??? Why is this in `.gitignore`?

Comment (freeranger) [inline .gitignore:18]: Because I obviously typed in the wrong window!
Updated and checked in

**Comment by tochman** [comment]
Will you be able to get this to work or should we move ahead with the other documentation tool for the time being?

**Comment by freeranger** [comment]
No, I don't know enough about rails and routing and I will get little or no time to code this week so go ahead with the other tool if you prefer, or live without having a test for this one, for now at least and add "Adding documentation tests" as a separate chore in the backlog.

**Comment by tochman** [comment]
It's not only about the tests. I can't access the documentation url

**Comment by freeranger** [comment]
What do you get when you go to localhost:3000/api?

**Comment by tochman** [comment]
skarmavbild 2016-02-22 kl 10 46 42

**Comment by sampritipanda** [comment]
Closing this in favor for Thomas's PR which uses apipie gem.

**Pull request closed without merging**